sets/sets.go

```diff
@@ -72,9 +72,29 @@ func (s *Set[T]) Remove(item T) bool {
 // Difference returns a set containing the elements of s that are not in x.
 func (s *Set[T]) Difference(x *Set[T]) *Set[T] {
 	result := New[T](max(0, s.Size()-x.Size()))
-	for i := range s.items {
-		if !x.Contains(i) {
-			result.items[i] = struct{}{}
+	for item := range s.items {
+		if !x.Contains(item) {
+			result.items[item] = struct{}{}
+		}
+	}
+	return result
+}
+
+// Intersection returns a set containing the elements that are both in s and x.
+func (s *Set[T]) Intersection(x *Set[T]) *Set[T] {
+	result := New[T](0)
+	// loop over the smaller set (thanks to https://github.com/deckarep/golang-set)
+	if s.Size() < x.Size() {
+		for item := range s.items {
+			if x.Contains(item) {
+				result.items[item] = struct{}{}
+			}
+		}
+	} else {
+		for item := range x.items {
+			if s.Contains(item) {
+				result.items[item] = struct{}{}
+			}
 		}
 	}
```

Comment:
Wouldn't it be better to just swap the variables and code the loop only once?
```suggestion
	if s.Size() > x.Size() {
		// Swap them to loop over the smallest set
		s, x = x, s
	}
	for item := range s.items {
		if x.Contains(item) {
			result.items[item] = struct{}{}
		}
	}
```

Reply:
This is swapping pointers (`s` is a pointer):
```
func (s *Set[T]) Intersection(...)
```

 but on the other hand, in Go, "assignment to the method receiver propagates only to callees but not to callers", so at the exit of the function, the contents of the set do remain the same. Said in another way, the code proposed is safe, although confusing.

Reply:
Just wanted to notice that GitHub itself decided that "marco-m-pix4d marked this conversation resolved".
I did nothing 🙄.

I added a second commit, that I thinks makes the code less confusing.